### PR TITLE
Add `StorageDataFilter`

### LINKED ISSUE #566: I suggest adding a new builtin filter
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
In my handlers I often use StorageDataFilter. It's similar as StateFilter and helps me to check storage data for current user and chat. Here are examples:

```python
dp = Dispatcher(bot)


# check if storage data for current user and chat has key 'regime' with value 'demo'
@dp.message_handler(storage_data={'regime': 'demo'})
async def test(msg: types.Message): ...


# check if has key 'regime' with value 'demo' and key 'game' with value 'football'
@dp.message_handler(storage_data={'regime': 'demo', 'game': 'football'})
async def test(msg: types.Message): ...


# check if has key 'game' with value 'football' or 'basketball'
# key 'game' with value ['football', 'basketball'] is also suitable
@dp.message_handler(storage_data={'game': ['football', 'basketball']})
async def test(msg: types.Message): ...


# check if has key 'regime' with value 'demo' and key 'game' with any value
@dp.message_handler(storage_data={'regime': 'demo', 'game': '*'})
async def test(msg: types.Message): ...


# check if has key 'game' with value '*' (or ['*'])
@dp.message_handler(storage={'game': ['*']})
async def test(msg: types.Message): ...
```

Here is code of filter:

```python
class StorageDataFilter(BoundFilter):
    """
    Check if all items matches the relevant items in the current storage data.
    """

    key = 'storage_data'
    ctx_storage_data = ContextVar('user_storage_data')

    def __init__(self, dispatcher, storage_data: dict):
        from aiogram import Dispatcher

        self.dispatcher: Dispatcher = dispatcher
        self.storage_data = storage_data

    @staticmethod
    def get_target(obj) -> typing.Tuple[Optional[int], Optional[int]]:
        if isinstance(obj, CallbackQuery):
            try:
                chat_id = obj.message.chat.id
            except AttributeError:
                chat_id = None
        else:
            try:
                chat_id = obj.chat.id
            except AttributeError:
                chat_id = None

        try:
            user_id = obj.from_user.id
        except AttributeError:
            user_id = None

        return chat_id, user_id

    async def get_current_storage_data(self, obj) -> Optional[dict]:
        try:
            return self.ctx_storage_data.get()
        except LookupError:
            chat_id, user_id = self.get_target(obj)

            if chat_id or user_id:
                storage_data = await self.dispatcher.storage.get_data(chat=chat_id, user=user_id)
                self.ctx_storage_data.set(storage_data)
                return storage_data

    async def check(self, obj) -> bool:
        current_storage_data = await self.get_current_storage_data(obj)

        if current_storage_data is None:
            return False

        for key, value in self.storage_data.items():
            if key not in current_storage_data:
                return False

            if value == '*':
                continue

            if isinstance(value, (list, tuple, set)):
                if current_storage_data[key] in value:
                    continue

            if current_storage_data[key] == value:
                continue

            return False
        return True
```

What about to include this filter in builtins filters?

### PR DESCRIPTION
**Pull request by LDmitriy7**
# Description

Fixes #566

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Tested all cases from issue #566

**Test Configuration**:
* Operating System: Windows 10
* Python version: 3.9

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: aiogram/dispatcher/filters/builtin.py [hunk 1]
```diff
@@ -757,3 +757,72 @@ def __init__(self, is_media_group: bool):
...
+    """
+
+    key = 'storage_data'
+    ctx_storage_data = ContextVar('user_storage_data')
+
+    def __init__(self, dispatcher, storage_data: dict):
+        from aiogram import Dispatcher
+
+        self.dispatcher: Dispatcher = dispatcher
+        self.storage_data = storage_data
```
Comment (Olegt0rr) [inline aiogram/dispatcher/filters/builtin.py:774]: Storage type hint set in arguments, but dispatcher type hint set inside the function.
Please make it similar way.

File: aiogram/dispatcher/filters/builtin.py [hunk 2]
```diff
@@ -757,3 +757,72 @@ def __init__(self, is_media_group: bool):
 
     async def check(self, message: types.Message) -> bool:
         return bool(getattr(message, "media_group_id")) is self.is_media_group
+
+
+class StorageDataFilter(BoundFilter):
+    """
+    Check if all items matches the relevant items in the current storage data.
+    """
```
Comment (Olegt0rr) [inline aiogram/dispatcher/filters/builtin.py:765]: Please, add examples to:
1. `examples` folder
2. `StorageDataFilter` docstrings

Also don't forget to add description of asterisk behaviour

**Review by Olegt0rr — CHANGES_REQUESTED** [review]

**Comment by JrooTJunior** [comment]
@LDmitriy7, FYI

Any news about requested changes?

**Comment by evgfilim1** [comment]
Closing as inactive. Feel free to comment here if you need or push your changes.

**Pull request closed without merging**